src/librbd/group/ListSnapshotsRequest.cc

```diff
@@ -0,0 +1,178 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "librbd/group/ListSnapshotsRequest.h"
+#include "include/ceph_assert.h"
+#include "common/dout.h"
+#include "common/errno.h"
+#include "common/ceph_context.h"
+#include "cls/rbd/cls_rbd_client.h"
+#include "librbd/Utils.h"
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::group::ListSnapshotsRequest: " << this \
+                           << " " << __func__ << ": "
+
+namespace librbd {
+namespace group {
+
+using util::create_rados_callback;
+
+namespace {
+
+static const uint32_t MAX_RETURN = 1024;
+
+} // anonymous namespace
+
+template <typename I>
+void ListSnapshotsRequest<I>::send() {
+  list_snap_orders();
+}
+
+template <typename I>
+void ListSnapshotsRequest<I>::list_snap_orders() {
+  if (!m_sort) {
+    list_snapshots();
+    return;
+  }
+
+  CephContext *cct = (CephContext *)m_group_io_ctx.cct();
+  ldout(cct, 10) << dendl;
+
+  librados::ObjectReadOperation op;
+  cls_client::group_snap_list_order_start(&op, m_start_after_order, MAX_RETURN);
+  auto comp = create_rados_callback< ListSnapshotsRequest<I>,
+                   &ListSnapshotsRequest<I>::handle_list_snap_orders>(this);
+  m_out_bl.clear();
+  int r = m_group_io_ctx.aio_operate(util::group_header_name(m_group_id), comp,
+                                     &op, &m_out_bl);
+  ceph_assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+void ListSnapshotsRequest<I>::handle_list_snap_orders(int r) {
+  CephContext *cct = (CephContext *)m_group_io_ctx.cct();
+  ldout(cct, 10) << "r=" << r << dendl;
+
+  std::map<std::string, uint64_t> snap_orders;
+  if (r == 0) {
+    auto iter = m_out_bl.cbegin();
+    r = librbd::cls_client::group_snap_list_order_finish(&iter, &snap_orders);
+  }
+
+  if (r < 0) {
```

Comment:
This needs to allow for `EOPNOTSUPP` if "fail if not ordered" behavior isn't requested.  Failing when running against older OSDs in that case would be bad.

Reply:
Done. Out of curiosity, what kind of errors are you thinking of? I thought maybe transient errors like a network issue could also be ignored if m_fail_if_not_sorted was false.

Reply:
Network errors would be handled (retried) by the messenger layer.  It could be something like a persistent decoding error caused by some silly mismatch between `cls_rbd_client.cc` and `cls_rbd.cc`, where the former encodes one set of fields while the latter expects to decode a slightly different one or an issue with bufferlist encoding in general.